src/backend/pipeline/cont_combiner.c

```diff
@@ -812,9 +786,18 @@ read_batch(ContQueryCombinerState *state, TupleBufferBatchReader *reader)
 
 		ExecStoreTuple(heap_copytuple(tbs->tuple->heaptup), state->slot, InvalidBuffer, false);
 		tuplestore_puttupleslot(state->batch, state->slot);
+		state->batch_hashes[count] = tbs->tuple->group_hash;
 
 		IncrementCQRead(1, tbs->size);
 		count++;
+
+		/*
+		 * This is fairly unlikely, but if it does happen we should
+		 * finish processing this batch immediately because we're storing 8-byte hashes
+		 * in memory alongside each of these tuples.
+		 */
+		if (!tuplestore_in_memory(state->batch))
```

Comment:
Ah you're right, I thought we were only using 32mb for the batch tuplestore. Even then it's within the realm of possible (10's of thousands of tuples x 10's of KB each), but at that point perf characteristics would be completely unpredictable anyways.

It still seems like an easy thing to avoid though, but I can revert this part if you'd like. I also updated the comment.